bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+   */
+  def onStackBind(bind: BindCommon, v: Variable) = {
+    bind.onStack = true
+  }
+  
+  def bindBindCommon(bind: BindCommon, sym: Symbol, right: Expression) = {
+		  walkExpr(right)
+		  // If we bind and it has not been used, put it on stack
+		  for (decl <- declVars.get(sym) if !decl.state.locked) {
+		    tagVariable(decl.v, BoundFirst)
+		    onStackBind(bind, decl.v)
+		  }
+  }
+  
+  override def walkStat(statement: Statement): Unit = statement match {
+    case bind @ BindStatement(left @ Variable(sym), right) =>
+      bindBindCommon(bind, sym, right)
+      
+    case CallStatement(callable, args) =>
+      walkExpr(callable)
+      args map walkExpr
+      for ((sym, decl) <- declVars if decl.state == Captured) {
+        tagVariable(decl.v, ToInstantiate)
+      }
+      
+    case IfStatement(cond, trueStat, elseStat) =>
+      walkExpr(cond)
+      conjugate(declVars, _.walkStat(trueStat), _.walkStat(elseStat))
+      
+    case MatchStatement(value, clauses, elseStatement) =>
+      walkExpr(value)
+      conjugate(declVars, (clauses.map {
+        clause =>
+          m: OnStackMarking =>
+           clause.guard foreach {g => m.walkExpr(g)}
+           m.walkStat(clause.body)
+      } :+ {m: OnStackMarking => m.walkStat(elseStatement)}) :_*)
+      
+    case TryStatement(body, exceptionVar, catchBody) =>
+      affectedVariables(null, { m =>
+        m.walkStat(body)
+        m.walkStat(catchBody)
+      }) foreach (tagVariable(_, ToInstantiate))
+      
+    case _ => super.walkStat(statement)
+  }
+  
+  override def walkExpr(expression: Expression): Unit = expression match {
+    case bind @ BindExpression(left @ Variable(sym), right) =>
+      bindBindCommon(bind, sym, right)
+      
+    case Variable(sym) =>
+      for (decl <- declVars.get(sym) if !decl.state.locked) {
+        tagVariable(decl.v, ToInstantiate)
+      }
+      
+    case ProcExpression(name, args, body, flags) =>
+      affectedVariables(null, { m =>
+        m.walkStat(body)
+      }) foreach (tagVariable(_, Captured))
+      
+    case IfExpression(cond, trueExpr, elseExpr) =>
+      walkExpr(cond)
+      conjugate(declVars, _.walkExpr(trueExpr), _.walkExpr(elseExpr))
+      
+    case MatchExpression(value, clauses, elseExpression) =>
+      walkExpr(value)
+      conjugate(declVars, (clauses.map { clause =>
+        m: OnStackMarking =>
+         clause.guard foreach {g => m.walkExpr(g)}
+         m.walkExpr(clause.body)
+      } :+ {m: OnStackMarking => m.walkExpr(elseExpression)}):_*)
+      
+    case TryExpression(body, exceptionVar, catchBody) =>
+      affectedVariables(null, { m =>
+        m.walkExpr(body)
+        m.walkExpr(catchBody)
+      }) foreach (tagVariable(_, ToInstantiate))
+      
+    case ShortCircuitBinaryOp(left, op, right) =>
+      affectedVariables(null, { m =>
+        m.walkExpr(left)
+        m.walkExpr(right)
+      }) foreach (tagVariable(_, ToInstantiate))
```

Comment:
The left could be walked outside of the affectedVariables call, because we are sure it will be executed. However, the right member could affect the state of a variable but not be executed. This is why, for safety, any variable affected there should be instantiated. This is kind of the same rationale as for try-catch blocks.
```oz
local A B
if B=true andthen A=true then
  skip
else
  A = true
end
```
If andthen was a simple binary operator, A could be put to true on stack. However, as the left member evaluates to false, A is not put to true, so the variable should be instantiated.

When ambiguous cases happen, the conservative approach of saying "if a variable is affected, instantiate it to let Oz do its job" seems pretty natural. It may sound inefficient, but since it only affects variables that are first seen or first bound in there, we cannot really do better.